File: zloop-skills/ai-wan-v16-orchestrator/lib/package-raw-cache.js
'use strict';
const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');
const crypto = require('node:crypto');
const { spawnSync } = require('node:child_process');

const CONTRACT_VERSION = 'ai-wan-v1.5.5-fetch';
const SCRIPTS = ['category_daily_avg','category_summary','category_fulfill_daily_avg','category_fulfill_summary','model_daily_avg','model_summary'];
const OPTIONAL_EMPTY_SCRIPTS = new Set(['category_fulfill_daily_avg','category_fulfill_summary']);
function ensureDir(d){fs.mkdirSync(d,{recursive:true});}
function sha256File(f){return crypto.createHash('sha256').update(fs.readFileSync(f)).digest('hex');}
function sha256Text(s){return crypto.createHash('sha256').update(String(s),'utf8').digest('hex');}
function writeJson(f,v){ensureDir(path.dirname(f));fs.writeFileSync(f,`${JSON.stringify(v,null,2)}\n`,'utf8');}
function parseArgs(argv=process.argv.slice(2)){const a={};for(let i=0;i<argv.length;i++){const t=argv[i];if(t==='--help'||t==='-h')a.help=true;else if(t.startsWith('--')){const k=t.slice(2).replace(/-([a-z])/g,(_,c)=>c.toUpperCase());const n=argv[i+1];if(!n||n.startsWith('--'))throw new Error(`Missing value for ${t}`);a[k]=n;i++;}else throw new Error(`Unknown argument: ${t}`);}return a;}
function run(cmd,args,opts={}){const r=spawnSync(cmd,args,{encoding:'utf8',...opts});if(r.status!==0)throw new Error(`${cmd} ${args.join(' ')} failed: ${r.stderr||r.stdout}`);return r;}
function zipDir(src,out,entries=['.']){run('sh',['-lc','command -v zip >/dev/null']);ensureDir(path.dirname(out));if(fs.existsSync(out))fs.rmSync(out,{force:true});run('zip',['-qr',path.resolve(out),...entries],{cwd:src});}
function splitCsvLine(line){const out=[];let cur='';let q=false;for(let i=0;i<line.length;i++){const ch=line[i];if(ch==='"'){if(q&&line[i+1]==='"'){cur+='"';i++;}else q=!q;}else if(ch===','&&!q){out.push(cur);cur='';}else cur+=ch;}out.push(cur);return out;}
function inspectCsv(file){const text=fs.readFileSync(file,'utf8').replace(/^\uFEFF/,'');const lines=text.split(/\r?\n/).filter(Boolean);return {row_count:Math.max(0,lines.length-1),column_count:lines.length?splitCsvLine(lines[0]).length:0,bytes:fs.statSync(file).size,sha256:sha256File(file)};}
function findInputFile(inputDir, script, ext){const exact=path.join(inputDir,`${script}.${ext}`);if(fs.existsSync(exact))return exact;const hit=fs.readdirSync(inputDir).find(f=>f===`${script}_${ext}`||f.startsWith(`${script}_`)&&f.endsWith(`.${ext}`));return hit?path.join(inputDir,hit):'';}
function packageRawCache(options={}){
 const runDt=options.runDt;if(!/^\d{4}-\d{2}-\d{2}$/.test(String(runDt||'')))throw new Error(`runDt must be YYYY-MM-DD, got ${runDt}`);
 const inputDir=path.resolve(options.inputDir||'.');const outDir=path.resolve(options.outDir||inputDir);const runId=options.runId||`fetch_${runDt}_${crypto.randomBytes(4).toString('hex')}`;ensureDir(outDir);
 const tmp=fs.mkdtempSync(path.join(os.tmpdir(),`ai-wan-fetch-${runDt}-`));
 try{ensureDir(path.join(tmp,'raw'));ensureDir(path.join(tmp,'sql'));const scripts={};const rawFiles=[];
  for(const script of SCRIPTS){const csv=findInputFile(inputDir,script,'csv');if(!csv)throw new Error(`missing raw csv for ${script} in ${inputDir}`);const sql=findInputFile(inputDir,script,'sql');const rawRel=`raw/${script}_${runDt}.csv`;const sqlRel=`sql/${script}_${runDt}.sql`;fs.copyFileSync(csv,path.join(tmp,rawRel));let sqlText=sql?fs.readFileSync(sql,'utf8'):`-- SQL text unavailable for ${script}; packaged by package-raw-cache.js\n`;fs.writeFileSync(path.join(tmp,sqlRel),sqlText,'utf8');const csvInfo=inspectCsv(path.join(tmp,rawRel));const emptyOptional=csvInfo.row_count===0&&OPTIONAL_EMPTY_SCRIPTS.has(script);scripts[script]={execute_id:options.executeId||'',status:csvInfo.row_count>0?'SUCCESS':(emptyOptional?'WARN':'FAILED'),row_count:csvInfo.row_count,column_count:csvInfo.column_count,bytes:csvInfo.bytes,sha256:csvInfo.sha256,raw_csv:rawRel,rendered_sql:sqlRel,rendered_sql_sha256:sha256Text(sqlText),started_at:options.startedAt||'',finished_at:options.finishedAt||new Date().toISOString(),error_summary:csvInfo.row_count>0?'':(emptyOptional?'empty csv accepted as fulfillment known gap':'empty csv')};rawFiles.push({script,path:rawRel,...csvInfo});}
  const values=Object.values(scripts);const status=values.every(s=>s.status==='SUCCESS')?'success':(values.some(s=>s.status==='FAILED')?'failed':'warn');
  const sqlStatus={contract_version:CONTRACT_VERSION,stage:'fetch',run_id:runId,run_dt:runDt,sql_scope:'all',status,scripts,generated_at:new Date().toISOString()};
  const knownGaps=options.knownGaps?String(options.knownGaps).split(',').filter(Boolean):[];for(const [script,info] of Object.entries(scripts)){if(info.status==='WARN'&&!knownGaps.includes(`${script}_empty`))knownGaps.push(`${script}_empty`);}
  const rawManifest={contract_version:CONTRACT_VERSION,stage:'fetch',run_id:runId,run_dt:runDt,target_month:runDt.slice(0,7),raw_files:rawFiles,known_gaps:knownGaps,generated_at:new Date().toISOString()};
  writeJson(path.join(tmp,`sql_status_${runDt}.json`),sqlStatus);writeJson(path.join(tmp,`raw_manifest_${runDt}.json`),rawManifest);
  writeJson(path.join(outDir,`sql_status_${runDt}.json`),sqlStatus);writeJson(path.join(outDir,`raw_manifest_${runDt}.json`),rawManifest);
  const rawCache=path.join(outDir,`raw_cache_${runDt}.zip`);zipDir(tmp,rawCache,['raw','sql',`sql_status_${runDt}.json`,`raw_manifest_${runDt}.json`]);
  const active={contract_version:CONTRACT_VERSION,stage:'fetch',status,run_id:runId,run_dt:runDt,target_month:runDt.slice(0,7),raw_cache:path.basename(rawCache),raw_cache_sha256:sha256File(rawCache),sha256:sha256File(rawCache),sql_status:`sql_status_${runDt}.json`,raw_manifest:`raw_manifest_${runDt}.json`,known_gaps:rawManifest.known_gaps,generated_at:new Date().toISOString()};
  writeJson(path.join(outDir,'active_fetch_manifest.json'),active);
  return {ok:status==='success'||status==='warn',active_manifest:active,raw_cache:rawCache,sql_status:sqlStatus,raw_manifest:rawManifest};
 }finally{if(options.keepWorkDir!==true)fs.rmSync(tmp,{recursive:true,force:true});}
}
module.exports={packageRawCache,parseArgs,inspectCsv,sha256File,SCRIPTS};

File: zloop-skills/ai-wan-v16-orchestrator/scripts/package_raw_cache.py
#!/usr/bin/env python3
"""Package AIWAN raw SQL exports into the fetch artifact contract.

This is the Python replacement for the historical Node package-raw-cache
wrapper.  It is stdlib-only so the remote Skill package does not depend on a
Node runtime for the Loop read stage.
"""

from __future__ import annotations

import argparse
import csv
import hashlib
import json
import os
import re
import secrets
import shutil
import sys
import tempfile
import zipfile
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable

if __package__ in {None, ""}:
    sys.path.insert(0, str(Path(__file__).resolve().parent))

from process_pipeline import BASE_SCRIPTS, FETCH_CONTRACT_VERSION, RAW_SCRIPTS  # noqa: E402


OPTIONAL_EMPTY_SCRIPTS = {"category_fulfill_daily_avg", "category_fulfill_summary"}


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def ensure_dir(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def sha256_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def sha256_text(text: str) -> str:
    return sha256_bytes(text.encode("utf-8"))


def sha256_file(path: Path) -> str:
    return sha256_bytes(path.read_bytes())


def write_json(path: Path, value: Any) -> None:
    ensure_dir(path.parent)
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def inspect_csv(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.reader(handle)
        try:
            headers = next(reader)
        except StopIteration:
            headers = []
            rows = 0
        else:
            rows = sum(1 for _ in reader)
    return {
        "row_count": rows,
        "column_count": len(headers),
        "bytes": path.stat().st_size,
        "sha256": sha256_file(path),
    }


def find_input_file(input_dir: Path, script: str, ext: str) -> Path | None:
    exact = input_dir / f"{script}.{ext}"
    if exact.exists():
        return exact
    hits = sorted(
        path for path in input_dir.iterdir()
        if path.is_file() and path.name.startswith(f"{script}_") and path.suffix == f".{ext}"
    )
    return hits[0] if hits else None


def resolve_script_scope(sql_scope: str | None = None, scripts: str | Iterable[str] | None = None) -> tuple[str, tuple[str, ...]]:
    scope = (sql_scope or "all").strip().lower()
    if scope not in {"all", "base"}:
        raise ValueError(f"sql_scope must be all or base, got {sql_scope}")

    if scripts is None:
        provided: list[str] = []
    elif isinstance(scripts, str):
        provided = [item.strip() for item in scripts.split(",") if item.strip()]
    else:
        provided = [str(item).strip() for item in scripts if str(item).strip()]

    active = tuple(provided or (BASE_SCRIPTS if scope == "base" else RAW_SCRIPTS))
    if len(set(active)) != len(active):
        raise ValueError(f"scripts contains duplicates: {','.join(active)}")

    expected = tuple(BASE_SCRIPTS if scope == "base" else RAW_SCRIPTS)
    unknown = [script for script in active if script not in RAW_SCRIPTS]
    missing = [script for script in expected if script not in active]
    extra = [script for script in active if script not in expected]
    if unknown or missing or extra:
        raise ValueError(
            f"scripts do not match sql_scope={scope}; "
            f"missing={','.join(missing) or '<none>'}; "
            f"extra={','.join(extra) or '<none>'}; "
            f"unknown={','.join(unknown) or '<none>'}"
        )
    return scope, active


def zip_dir(source_dir: Path, out_file: Path, entries: Iterable[str]) -> None:
    ensure_dir(out_file.parent)
    if out_file.exists():
        out_file.unlink()
    with zipfile.ZipFile(out_file, "w", compression=zipfile.ZIP_DEFLATED) as archive:
        for entry in entries:
            root = source_dir / entry
            if root.is_dir():
                for path in sorted(root.rglob("*")):
                    if path.is_file():
                        archive.write(path, path.relative_to(source_dir).as_posix())
            elif root.is_file():
                archive.write(root, root.relative_to(source_dir).as_posix())


def package_raw_cache(
    *,
    run_dt: str,
    input_dir: str | Path = ".",
    out_dir: str | Path | None = None,
    run_id: str | None = None,
    known_gaps: str | Iterable[str] | None = None,
    sql_scope: str | None = None,
    scripts: str | Iterable[str] | None = None,
    keep_work_dir: bool = False,
) -> dict[str, Any]:
    if not run_dt or not isinstance(run_dt, str) or not re.match(r"^\d{4}-\d{2}-\d{2}$", run_dt):
        raise ValueError(f"run_dt must be YYYY-MM-DD, got {run_dt}")

    input_path = Path(input_dir).resolve()
    output_path = Path(out_dir or input_path).resolve()
    fetch_id = run_id or f"fetch_{run_dt}_{secrets.token_hex(4)}"
    scope, active_scripts = resolve_script_scope(sql_scope, scripts)
    ensure_dir(output_path)
    work_dir = Path(tempfile.mkdtemp(prefix=f"ai-wan-fetch-{run_dt}-"))
    try:
        ensure_dir(work_dir / "raw")
        ensure_dir(work_dir / "sql")
        script_status: dict[str, Any] = {}
        raw_files: list[dict[str, Any]] = []

        for script in active_scripts:
            csv_file = find_input_file(input_path, script, "csv")
            if csv_file is None:
                raise FileNotFoundError(f"missing raw csv for {script} in {input_path}")
            sql_file = find_input_file(input_path, script, "sql")
            raw_rel = Path("raw") / f"{script}_{run_dt}.csv"
            sql_rel = Path("sql") / f"{script}_{run_dt}.sql"
            shutil.copyfile(csv_file, work_dir / raw_rel)
            sql_text = (
                sql_file.read_text(encoding="utf-8")
                if sql_file is not None
                else f"-- SQL text unavailable for {script}; packaged by package_raw_cache.py\n"
            )
            (work_dir / sql_rel).write_text(sql_text, encoding="utf-8")

            csv_info = inspect_csv(work_dir / raw_rel)
            empty_optional = csv_info["row_count"] == 0 and script in OPTIONAL_EMPTY_SCRIPTS
            status = "SUCCESS" if csv_info["row_count"] > 0 else ("WARN" if empty_optional else "FAILED")
            script_status[script] = {
                "execute_id": "",
                "status": status,
                "row_count": csv_info["row_count"],
                "column_count": csv_info["column_count"],
                "bytes": csv_info["bytes"],
                "sha256": csv_info["sha256"],
                "raw_csv": raw_rel.as_posix(),
                "rendered_sql": sql_rel.as_posix(),
                "rendered_sql_sha256": sha256_text(sql_text),
                "started_at": "",
                "finished_at": now_iso(),
                "error_summary": "" if status == "SUCCESS" else ("empty csv accepted as fulfillment known gap" if empty_optional else "empty csv"),
            }
            raw_files.append({"script": script, "path": raw_rel.as_posix(), **csv_info})

        values = list(script_status.values())
        status = "success" if all(item["status"] == "SUCCESS" for item in values) else (
            "failed" if any(item["status"] == "FAILED" for item in values) else "warn"
        )
        if known_gaps is None:
            gaps: list[str] = []
        elif isinstance(known_gaps, str):
            gaps = [item for item in known_gaps.split(",") if item]
        else:
            gaps = [str(item) for item in known_gaps if str(item)]
        for script, info in script_status.items():
            gap = f"{script}_empty"
            if info["status"] == "WARN" and gap not in gaps:
                gaps.append(gap)

        sql_status = {
            "contract_version": FETCH_CONTRACT_VERSION,
            "stage": "fetch",
            "run_id": fetch_id,
            "run_dt": run_dt,
            "sql_scope": scope,
            "active_scripts": list(active_scripts),
            "status": status,
            "scripts": script_status,
            "generated_at": now_iso(),
        }
        raw_manifest = {
            "contract_version": FETCH_CONTRACT_VERSION,
            "stage": "fetch",
            "run_id": fetch_id,
            "run_dt": run_dt,
            "target_month": run_dt[:7],
            "sql_scope": scope,
            "scripts": list(active_scripts),
            "raw_files": raw_files,
            "known_gaps": gaps,
            "generated_at": now_iso(),
        }

        write_json(work_dir / f"sql_status_{run_dt}.json", sql_status)
        write_json(work_dir / f"raw_manifest_{run_dt}.json", raw_manifest)
        write_json(output_path / f"sql_status_{run_dt}.json", sql_status)
        write_json(output_path / f"raw_manifest_{run_dt}.json", raw_manifest)

        raw_cache = output_path / f"raw_cache_{run_dt}.zip"
        zip_dir(work_dir, raw_cache, ("raw", "sql", f"sql_status_{run_dt}.json", f"raw_manifest_{run_dt}.json"))
        active = {
            "contract_version": FETCH_CONTRACT_VERSION,
            "stage": "fetch",
            "status": status,
            "run_id": fetch_id,
            "run_dt": run_dt,
            "target_month": run_dt[:7],
            "sql_scope": scope,
            "scripts": list(active_scripts),
            "raw_cache": raw_cache.name,
            "raw_cache_sha256": sha256_file(raw_cache),
            "sha256": sha256_file(raw_cache),
            "sql_status": f"sql_status_{run_dt}.json",
            "raw_manifest": f"raw_manifest_{run_dt}.json",
            "known_gaps": gaps,
            "generated_at": now_iso(),
        }
        write_json(output_path / "active_fetch_manifest.json", active)
        return {"ok": status in {"success", "warn"}, "active_manifest": active, "raw_cache": str(raw_cache), "sql_status": sql_status, "raw_manifest": raw_manifest}
    finally:
        if not keep_work_dir:
            shutil.rmtree(work_dir, ignore_errors=True)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Package selected xinghe raw CSVs into raw_cache/sql_status/raw_manifest artifacts."
    )
    parser.add_argument("--run-dt", default=os.environ.get("RUN_DT"))
    parser.add_argument("--run-id", default=os.environ.get("RUN_ID"))
    parser.add_argument("--input-dir", default=os.environ.get("INPUT_DIR", "."))
    parser.add_argument("--out-dir", default=os.environ.get("OUT_DIR", "."))
    parser.add_argument("--known-gaps", default=os.environ.get("KNOWN_GAPS"))
    parser.add_argument("--sql-scope", default=os.environ.get("SQL_SCOPE"))
    parser.add_argument("--scripts", default=os.environ.get("SCRIPTS"))
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if not args.run_dt:
        print("--run-dt is required", file=sys.stderr)
        return 2
    try:
        result = package_raw_cache(
            run_dt=args.run_dt,
            run_id=args.run_id,
            input_dir=args.input_dir,
            out_dir=args.out_dir,
            known_gaps=args.known_gaps,
            sql_scope=args.sql_scope,
            scripts=args.scripts,
        )
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0 if result.get("ok") else 1
    except Exception as exc:  # pragma: no cover - defensive CLI boundary
        print(f"{type(exc).__name__}: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
